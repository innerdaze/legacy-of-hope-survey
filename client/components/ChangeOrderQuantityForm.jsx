import React, { Component } from 'react'
import Layer from 'grommet/components/Layer'
import Form from 'grommet/components/Form'
import Header from 'grommet/components/Header'
import Heading from 'grommet/components/Heading'
import Footer from 'grommet/components/Footer'
import Button from 'grommet/components/Button'
import Anchor from 'grommet/components/Anchor'
import FormField from 'grommet/components/FormField'
import NumberInput from 'grommet/components/NumberInput'
import CloseIcon from 'grommet/components/icons/base/Close'
import NextLinkIcon from 'grommet/components/icons/base/LinkNext'

class ChangeOrderQuantityForm extends Component {

  constructor(props) {
    super(props)

    this.state = {
      quantity: null
    }

    this.onNumberInputChange = this.onNumberInputChange.bind(this)
    this.onSubmit = this.onSubmit.bind(this)
    this.onCancel = this.onCancel.bind(this)
  }

  componentDidMount() {
    this.setState({
      quantity: this.props.order.Qty
    })
  }

  onNumberInputChange(e) {
    this.setState({
      quantity: parseInt(e.target.value)
    })
  }

  onSubmit(e) {
    e.preventDefault()

    this.props.changeOrderQuantity(this.props.order.Barcode, this.state.quantity)

    this.props.onSubmit()
  }

  onCancel(e) {
    e.preventDefault()

    this.props.onCancel()
  }

  render() {
    return (
      <Layer>
        <Form>
          <Header>
            <Heading>
              Change Order Quantity
            </Heading>
          </Header>
          <FormField>
            <NumberInput
              min={1}
              value={this.state.quantity}
              onChange={this.onNumberInputChange}/>
          </FormField>
          <Footer
            pad={{'vertical': 'medium'}}
            justify='between'>
            <Anchor
              label='Cancel'
              icon={<CloseIcon/>}
              onClick={this.onCancel}/>
            <Anchor
              primary={true}
              label='Submit'
              icon={<NextLinkIcon/>}
              onClick={this.onSubmit}/>
          </Footer>
        </Form>
      </Layer>
    )
  }
}

ChangeOrderQuantityForm.propTypes = {
  order: React.PropTypes.object.isRequired,
  onSubmit: React.PropTypes.func.isRequired,
  onCancel: React.PropTypes.func.isRequired
}

ChangeOrderQuantityForm.propDefaults = {
  order: null,
  onChangeOrderQuantitySubmit: Function.prototype,
  onChangeOrderQuantityChange: Function.prototype,
}

export default ChangeOrderQuantityForm
