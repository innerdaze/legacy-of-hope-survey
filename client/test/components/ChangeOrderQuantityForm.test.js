import React from 'react'
import ChangeOrderQuantityForm from '../../components/ChangeOrderQuantityForm'

let wrapper, order, handleSubmit, handleCancel

describe('Testing on ChangeOrderQuantityForm...', () => {
  beforeEach(() => {
    order = {}
    handleSubmit = jest.fn()
    handleCancel = jest.fn()

    wrapper = shallow(
      <ChangeOrderQuantityForm
        order={order}
        handleSubmit={handleSubmit}
        handleCancel={handleCancel}
      />
    )
  })

  describe('ChangeOrderQuantityForm layout', () => {
    test('Expect ChangeOrderQuantityForm to have a Form with children comp, Header, input and Footer', () => {
      expect(wrapper.find('Form')).toHaveLength(1)
      expect(
        wrapper
          .find('Form')
          .find('FormField')
          .find('input')
      ).toHaveLength(1)
      expect(wrapper.find('Form').find('Footer')).toHaveLength(1)
      expect(wrapper).toMatchSnapshot()
    })

    test('Expect Header to have a Heading comp in it', () => {
      expect(wrapper.find('Header').find('Heading')).toHaveLength(1)
      expect(wrapper).toMatchSnapshot()
    })

    test('Expect to show Label if order prop is valid object', () => {
      order = {
        ProductName: 'Test Label',
        Qty: 2,
        _id: 1
      }
      handleSubmit = jest.fn()
      handleCancel = jest.fn()

      wrapper = shallow(
        <ChangeOrderQuantityForm
          order={order}
          handleSubmit={handleSubmit}
          handleCancel={handleCancel}
        />
      )

      expect(wrapper.find('Box').find('Label')).toHaveLength(1)
      expect(
        wrapper
          .find('Box')
          .find('Label')
          .html()
      ).toContain('Test Label')
      expect(wrapper).toMatchSnapshot()
    })

    test('Expect Footer to have 2 Anchor child comps in it', () => {
      expect(wrapper.find('Footer').find('Anchor')).toHaveLength(2)
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Test on prop', () => {
    test('Expect prop function handleSubmit to be called if Form submit or Anchor Update is clicked', () => {
      wrapper.find('Anchor[label="Update"]').simulate('click', {
        preventDefault() {}
      })
      expect(handleSubmit).toBeCalled()

      wrapper.find('Form').simulate('submit', {
        preventDefault() {}
      })
      expect(handleSubmit.mock.calls).toHaveLength(2)

      expect(wrapper).toMatchSnapshot()
    })

    test('Expect prop function handleCancel to be called if Anchor Cancel is clicked', () => {
      wrapper.find('Anchor[label="Cancel"]').simulate('click', {
        preventDefault() {}
      })
      expect(handleCancel).toBeCalled()
      expect(wrapper).toMatchSnapshot()
    })
  })

  describe('Test on state', () => {
    test('Expect quantity state to equal order prop Qty value', () => {
      order = {
        ProductName: 'Test Label',
        Qty: 2,
        _id: 1
      }
      wrapper = shallow(
        <ChangeOrderQuantityForm
          order={order}
          handleSubmit={handleSubmit}
          handleCancel={handleCancel}
        />
      )
      expect(wrapper.state('quantity')).toEqual(order.Qty)
      expect(wrapper).toMatchSnapshot()
    })

    test('Expect quantity state to change if input field is changed', () => {
      wrapper.find('input').prop('onChange', {
        preventDefault() {}
      })({
        target: {
          value: 20
        }
      })

      expect(wrapper.state('quantity')).toEqual(20)
      expect(wrapper).toMatchSnapshot()
    })
  })
})
