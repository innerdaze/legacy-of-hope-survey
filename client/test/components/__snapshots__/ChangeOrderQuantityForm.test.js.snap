// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing on ChangeOrderQuantityForm... ChangeOrderQuantityForm layout Expect ChangeOrderQuantityForm to have a Form with children comp, Header, input and Footer 1`] = `
<Layer
  align="center"
>
  <Form
    compact={false}
    fill={false}
    onSubmit={[Function]}
    pad="none"
  >
    <Header
      align="center"
      direction="row"
      pad={
        Object {
          "between": "small",
          "horizontal": "none",
          "vertical": "none",
        }
      }
      responsive={false}
    >
      <Heading
        tag="h1"
      >
        Set Quantity
      </Heading>
    </Header>
    <FormField
      size="medium"
      strong={false}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter quantity"
        type="number"
      />
    </FormField>
    <Footer
      align="center"
      direction="row"
      justify="between"
      pad={
        Object {
          "vertical": "medium",
        }
      }
      responsive={false}
    >
      <Anchor
        animateIcon={true}
        icon={
          <Close
            responsive={true}
          />
        }
        label="Cancel"
        method="push"
        onClick={[Function]}
        tag="a"
      />
      <Anchor
        animateIcon={true}
        icon={
          <LinkNext
            responsive={true}
          />
        }
        label="Update"
        method="push"
        onClick={[Function]}
        primary={true}
        tag="a"
      />
    </Footer>
  </Form>
</Layer>
`;

exports[`Testing on ChangeOrderQuantityForm... ChangeOrderQuantityForm layout Expect Footer to have 2 Anchor child comps in it 1`] = `
<Layer
  align="center"
>
  <Form
    compact={false}
    fill={false}
    onSubmit={[Function]}
    pad="none"
  >
    <Header
      align="center"
      direction="row"
      pad={
        Object {
          "between": "small",
          "horizontal": "none",
          "vertical": "none",
        }
      }
      responsive={false}
    >
      <Heading
        tag="h1"
      >
        Set Quantity
      </Heading>
    </Header>
    <FormField
      size="medium"
      strong={false}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter quantity"
        type="number"
      />
    </FormField>
    <Footer
      align="center"
      direction="row"
      justify="between"
      pad={
        Object {
          "vertical": "medium",
        }
      }
      responsive={false}
    >
      <Anchor
        animateIcon={true}
        icon={
          <Close
            responsive={true}
          />
        }
        label="Cancel"
        method="push"
        onClick={[Function]}
        tag="a"
      />
      <Anchor
        animateIcon={true}
        icon={
          <LinkNext
            responsive={true}
          />
        }
        label="Update"
        method="push"
        onClick={[Function]}
        primary={true}
        tag="a"
      />
    </Footer>
  </Form>
</Layer>
`;

exports[`Testing on ChangeOrderQuantityForm... ChangeOrderQuantityForm layout Expect Header to have a Heading comp in it 1`] = `
<Layer
  align="center"
>
  <Form
    compact={false}
    fill={false}
    onSubmit={[Function]}
    pad="none"
  >
    <Header
      align="center"
      direction="row"
      pad={
        Object {
          "between": "small",
          "horizontal": "none",
          "vertical": "none",
        }
      }
      responsive={false}
    >
      <Heading
        tag="h1"
      >
        Set Quantity
      </Heading>
    </Header>
    <FormField
      size="medium"
      strong={false}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter quantity"
        type="number"
      />
    </FormField>
    <Footer
      align="center"
      direction="row"
      justify="between"
      pad={
        Object {
          "vertical": "medium",
        }
      }
      responsive={false}
    >
      <Anchor
        animateIcon={true}
        icon={
          <Close
            responsive={true}
          />
        }
        label="Cancel"
        method="push"
        onClick={[Function]}
        tag="a"
      />
      <Anchor
        animateIcon={true}
        icon={
          <LinkNext
            responsive={true}
          />
        }
        label="Update"
        method="push"
        onClick={[Function]}
        primary={true}
        tag="a"
      />
    </Footer>
  </Form>
</Layer>
`;

exports[`Testing on ChangeOrderQuantityForm... ChangeOrderQuantityForm layout Expect to show Label if order prop is valid object 1`] = `
<Layer
  align="center"
>
  <Form
    compact={false}
    fill={false}
    onSubmit={[Function]}
    pad="none"
  >
    <Header
      align="center"
      direction="row"
      pad={
        Object {
          "between": "small",
          "horizontal": "none",
          "vertical": "none",
        }
      }
      responsive={false}
    >
      <Heading
        tag="h1"
      >
        Set Quantity
      </Heading>
    </Header>
    <Box
      announce={false}
      direction="column"
      focusable={true}
      pad="none"
      responsive={true}
      tag="div"
    >
      <Label
        size="medium"
      >
        Test Label
      </Label>
    </Box>
    <FormField
      size="medium"
      strong={false}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter quantity"
        type="number"
        value={2}
      />
    </FormField>
    <Footer
      align="center"
      direction="row"
      justify="between"
      pad={
        Object {
          "vertical": "medium",
        }
      }
      responsive={false}
    >
      <Anchor
        animateIcon={true}
        icon={
          <Close
            responsive={true}
          />
        }
        label="Cancel"
        method="push"
        onClick={[Function]}
        tag="a"
      />
      <Anchor
        animateIcon={true}
        icon={
          <LinkNext
            responsive={true}
          />
        }
        label="Update"
        method="push"
        onClick={[Function]}
        primary={true}
        tag="a"
      />
    </Footer>
  </Form>
</Layer>
`;

exports[`Testing on ChangeOrderQuantityForm... Test on prop Expect prop function handleCancel to be called if Anchor Cancel is clicked 1`] = `
<Layer
  align="center"
>
  <Form
    compact={false}
    fill={false}
    onSubmit={[Function]}
    pad="none"
  >
    <Header
      align="center"
      direction="row"
      pad={
        Object {
          "between": "small",
          "horizontal": "none",
          "vertical": "none",
        }
      }
      responsive={false}
    >
      <Heading
        tag="h1"
      >
        Set Quantity
      </Heading>
    </Header>
    <FormField
      size="medium"
      strong={false}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter quantity"
        type="number"
      />
    </FormField>
    <Footer
      align="center"
      direction="row"
      justify="between"
      pad={
        Object {
          "vertical": "medium",
        }
      }
      responsive={false}
    >
      <Anchor
        animateIcon={true}
        icon={
          <Close
            responsive={true}
          />
        }
        label="Cancel"
        method="push"
        onClick={[Function]}
        tag="a"
      />
      <Anchor
        animateIcon={true}
        icon={
          <LinkNext
            responsive={true}
          />
        }
        label="Update"
        method="push"
        onClick={[Function]}
        primary={true}
        tag="a"
      />
    </Footer>
  </Form>
</Layer>
`;

exports[`Testing on ChangeOrderQuantityForm... Test on prop Expect prop function handleSubmit to be called if Form submit or Anchor Update is clicked 1`] = `
<Layer
  align="center"
>
  <Form
    compact={false}
    fill={false}
    onSubmit={[Function]}
    pad="none"
  >
    <Header
      align="center"
      direction="row"
      pad={
        Object {
          "between": "small",
          "horizontal": "none",
          "vertical": "none",
        }
      }
      responsive={false}
    >
      <Heading
        tag="h1"
      >
        Set Quantity
      </Heading>
    </Header>
    <FormField
      size="medium"
      strong={false}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter quantity"
        type="number"
      />
    </FormField>
    <Footer
      align="center"
      direction="row"
      justify="between"
      pad={
        Object {
          "vertical": "medium",
        }
      }
      responsive={false}
    >
      <Anchor
        animateIcon={true}
        icon={
          <Close
            responsive={true}
          />
        }
        label="Cancel"
        method="push"
        onClick={[Function]}
        tag="a"
      />
      <Anchor
        animateIcon={true}
        icon={
          <LinkNext
            responsive={true}
          />
        }
        label="Update"
        method="push"
        onClick={[Function]}
        primary={true}
        tag="a"
      />
    </Footer>
  </Form>
</Layer>
`;

exports[`Testing on ChangeOrderQuantityForm... Test on state Expect quantity state to change if input field is changed 1`] = `
<Layer
  align="center"
>
  <Form
    compact={false}
    fill={false}
    onSubmit={[Function]}
    pad="none"
  >
    <Header
      align="center"
      direction="row"
      pad={
        Object {
          "between": "small",
          "horizontal": "none",
          "vertical": "none",
        }
      }
      responsive={false}
    >
      <Heading
        tag="h1"
      >
        Set Quantity
      </Heading>
    </Header>
    <FormField
      size="medium"
      strong={false}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter quantity"
        type="number"
      />
    </FormField>
    <Footer
      align="center"
      direction="row"
      justify="between"
      pad={
        Object {
          "vertical": "medium",
        }
      }
      responsive={false}
    >
      <Anchor
        animateIcon={true}
        icon={
          <Close
            responsive={true}
          />
        }
        label="Cancel"
        method="push"
        onClick={[Function]}
        tag="a"
      />
      <Anchor
        animateIcon={true}
        icon={
          <LinkNext
            responsive={true}
          />
        }
        label="Update"
        method="push"
        onClick={[Function]}
        primary={true}
        tag="a"
      />
    </Footer>
  </Form>
</Layer>
`;

exports[`Testing on ChangeOrderQuantityForm... Test on state Expect quantity state to equal order prop Qty value 1`] = `
<Layer
  align="center"
>
  <Form
    compact={false}
    fill={false}
    onSubmit={[Function]}
    pad="none"
  >
    <Header
      align="center"
      direction="row"
      pad={
        Object {
          "between": "small",
          "horizontal": "none",
          "vertical": "none",
        }
      }
      responsive={false}
    >
      <Heading
        tag="h1"
      >
        Set Quantity
      </Heading>
    </Header>
    <Box
      announce={false}
      direction="column"
      focusable={true}
      pad="none"
      responsive={true}
      tag="div"
    >
      <Label
        size="medium"
      >
        Test Label
      </Label>
    </Box>
    <FormField
      size="medium"
      strong={false}
    >
      <input
        autoFocus={true}
        onChange={[Function]}
        placeholder="Enter quantity"
        type="number"
        value={2}
      />
    </FormField>
    <Footer
      align="center"
      direction="row"
      justify="between"
      pad={
        Object {
          "vertical": "medium",
        }
      }
      responsive={false}
    >
      <Anchor
        animateIcon={true}
        icon={
          <Close
            responsive={true}
          />
        }
        label="Cancel"
        method="push"
        onClick={[Function]}
        tag="a"
      />
      <Anchor
        animateIcon={true}
        icon={
          <LinkNext
            responsive={true}
          />
        }
        label="Update"
        method="push"
        onClick={[Function]}
        primary={true}
        tag="a"
      />
    </Footer>
  </Form>
</Layer>
`;
